File: platform-common/src/main/java/org/sweetchips/platform/common/PathUnit.java
package org.sweetchips.platform.common;

import org.sweetchips.utility.AsyncUtil;
import org.sweetchips.utility.FilesUtil;

import java.nio.file.Path;
import java.util.Collections;
import java.util.List;
import java.util.function.BiFunction;

public final class PathUnit extends AbstractUnit {

    private final List<BiFunction<Path, Path, AbstractUnit>> mPrepare;
    private final List<BiFunction<Path, Path, AbstractUnit>> mTransform;

    public PathUnit(Path input, Path output,
                    List<BiFunction<Path, Path, AbstractUnit>> prepare,
                    List<BiFunction<Path, Path, AbstractUnit>> transform) {
        super(input, output);
        mPrepare = prepare != null ? prepare : Collections.emptyList();
        mTransform = transform != null ? transform : Collections.emptyList();
    }

    @Override
    protected final void onPrepare() {
        if (mPrepare.isEmpty()) {
            return;
        }
        AsyncUtil.with(FilesUtil.list(getInput()))
                .forkJoin(it -> {
                    if (!FilesUtil.exists(it)) {
                        return;
                    }
                    Path path = FilesUtil.lookupPathFromTo(it, getInput(), getOutput());
                    AbstractUnit unit = null;
                    for (BiFunction<Path, Path, AbstractUnit> filter : mPrepare) {
                        if ((unit = filter.apply(it, path)) != null) {
                            break;
                        }
                    }
                    if (unit == null) {
                        if (FilesUtil.isDirectory(it)) {
                            unit = new PathUnit(it, path, mPrepare, mTransform);
                        }
                    }
                    if (unit != null) {
                        unit.onPrepare();
                    }
                });
    }

    @Override
    protected final void onTransform() {
        AsyncUtil.with(FilesUtil.list(getInput()))
                .forkJoin(it -> {
                    if (!FilesUtil.exists(it)) {
                        return;
                    }
                    Path path = FilesUtil.lookupPathFromTo(it, getInput(), getOutput());
                    AbstractUnit abstractUnit = null;
                    for (BiFunction<Path, Path, AbstractUnit> filter : mTransform) {
                        if ((abstractUnit = filter.apply(it, path)) != null) {
                            break;
                        }
                    }
                    if (abstractUnit == null) {
                        if (FilesUtil.isDirectory(it)) {
                            abstractUnit = new PathUnit(it, path, mPrepare, mTransform);
                        } else {
                            abstractUnit = new FileUnit(it, path, null, null);
                        }
                    }
                    abstractUnit.onTransform();
                });
    }
}

File: platform-common/src/main/kotlin/org/sweetchips/platform/common/PathUnit.kt
package org.sweetchips.platform.common

import kotlinx.coroutines.launch
import kotlinx.coroutines.supervisorScope
import org.sweetchips.utility.FilesUtil
import java.nio.file.Path
import java.util.function.BiFunction
import kotlin.streams.toList

class PathUnit(
    input: Path,
    output: Path,
    private val prepare: List<BiFunction<Path, Path, IUnit?>>?,
    private val transform: List<BiFunction<Path, Path, IUnit?>>?,
) : AbstractUnit(input, output) {

    override suspend fun onPrepare() {
        if (prepare?.isEmpty() != false) {
            return
        }
        supervisorScope {
            FilesUtil.list(input).toList().forEach {
                if (!FilesUtil.exists(it)) {
                    return@forEach
                }
                val path = FilesUtil.lookupPathFromTo(it, input, output)
                var unit: IUnit? = null
                for (filter in prepare) {
                    unit = filter.apply(it, path)
                    if (unit != null) {
                        break
                    }
                }
                if (unit == null && FilesUtil.isDirectory(it)) {
                    unit = PathUnit(it, path, prepare, transform)
                }
                launch {
                    unit?.onPrepare()
                }
            }
        }
    }

    override suspend fun onTransform() {
        if (transform?.isEmpty() != false) {
            return
        }
        supervisorScope {
            FilesUtil.list(input).toList().forEach {
                if (!FilesUtil.exists(it)) {
                    return@forEach
                }
                val path = FilesUtil.lookupPathFromTo(it, input, output)
                var unit: IUnit? = null
                for (filter in transform) {
                    unit = filter.apply(it, path)
                    if (unit != null) {
                        break
                    }
                }
                if (unit == null) {
                    if (FilesUtil.isDirectory(it)) {
                        unit = PathUnit(it, path, prepare, transform)
                    } else {
                        unit = FileUnit(it, path, null, null)
                    }
                }
                launch {
                    unit.onTransform()
                }
            }
        }
    }
}